File: src/agentiq/web/static/reducer.js
// JS mirror of agentiq/replay/reducer.py — the SINGLE scene projection.
// Must fold identically to the Python reducer (NFR5); any change there lands
// here too. Kept dependency-free so it can run in the browser and under Node
// for the parity fixture test.

export const Zone = {
  FLOOR: "floor",
  LIBRARY: "library",
  DESK: "desk",
  SUBAGENTS: "subagents",
  PATH: "path",
};

export const Status = {
  IDLE: "idle",
  READING: "reading",
  THINKING: "thinking",
  DELEGATING: "delegating",
  WORKING: "working",
  FAILED: "failed",
  DONE: "done",
};

export function initialState() {
  return {
    current_seq: -1,
    run_status: "pending",
    caption: "",
    agents: {}, // agent_id -> {agent_id, role, parent_id, status, zone, team}
    paths: [], // [from_id, to_id][]
  };
}

export function reduce(state, event) {
  const agents = {};
  for (const [aid, a] of Object.entries(state.agents)) agents[aid] = { ...a };
  const paths = state.paths.slice();
  let runStatus = state.run_status;
  let caption = state.caption;
  const aid = event.agent_id;
  const p = event.payload ?? {};

  switch (event.type) {
    case "run.started":
      runStatus = "running";
      caption = `run started: ${p.goal ?? ""}`;
      break;
    case "run.completed":
      runStatus = p.status ?? "completed";
      caption = `run ${runStatus}`;
      break;
    case "run.aborted":
      runStatus = "aborted";
      caption = `run aborted: ${p.reason ?? ""}`;
      break;
    case "agent.spawned": {
      const key = aid ?? "";
      const parentId = p.parent_id ?? null;
      agents[key] = {
        agent_id: key,
        role: p.role ?? null,
        parent_id: parentId,
        status: Status.IDLE,
        zone: parentId ? Zone.DESK : Zone.FLOOR,
        team: p.team ?? null,
      };
      caption = `spawned ${key}` + (p.role ? ` (${p.role})` : "");
      break;
    }
    case "task.delegated": {
      const toAgent = p.to_agent ?? "";
      paths.push([aid ?? "", toAgent]);
      if (aid && agents[aid]) {
        agents[aid].status = Status.DELEGATING;
        agents[aid].zone = Zone.PATH;
      }
      if (agents[toAgent]) agents[toAgent].status = Status.WORKING;
      caption = `${aid} -> ${toAgent}: ${p.task ?? ""}`;
      break;
    }
    case "vault.read":
      if (aid && agents[aid]) {
        agents[aid].zone = Zone.LIBRARY;
        agents[aid].status = Status.READING;
      }
      caption = `${aid} reads vault: ${p.ref ?? ""}`;
      break;
    case "decision.pending":
      caption = `decision pending: ${p.prompt ?? ""}`;
      break;
    case "decision.resolved":
      caption = `decision resolved: ${p.choice ?? ""}`;
      break;
    case "agent.failed":
      if (aid && agents[aid]) agents[aid].status = Status.FAILED;
      caption = `${aid} failed: ${p.cause ?? ""}`;
      break;
    default:
      // Valid but unhandled type: advance without crashing (forward-compat).
      caption = event.type;
  }

  return {
    current_seq: event.seq,
    run_status: runStatus,
    caption,
    agents,
    paths,
  };
}

export function reduceAll(events, initial = null) {
  let state = initial ?? initialState();
  for (const event of events) state = reduce(state, event);
  return state;
}

// Precompute one state per event index — O(1) seeking, like ReplayController.
export function statesPerIndex(events) {
  const states = [];
  let state = initialState();
  for (const event of events) {
    state = reduce(state, event);
    states.push(state);
  }
  return states;
}

export function isDecision(eventType) {
  return eventType.startsWith("decision.");
}

export function isFailure(eventType) {
  return eventType === "agent.failed" || eventType === "run.aborted";
}

File: src/agentiq/replay/reducer.py
"""The single pure reducer: events → SceneState.

`reduce(state, event)` returns a NEW state and never mutates the input, so the
transport (story 2.3) can cache and step states freely. Both replay and the live
scene fold events through this one function, guaranteeing identical output for
identical input (NFR5).
"""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import replace

from agentiq.events.models import Event
from agentiq.replay.scene_state import (
    AgentState,
    SceneState,
    Status,
    Zone,
    initial_state,
)


def reduce(state: SceneState, event: Event) -> SceneState:
    """Fold one event into the scene, returning a new (never-mutated) state."""
    agents = {aid: replace(a) for aid, a in state.agents.items()}
    paths = list(state.paths)
    run_status = state.run_status
    caption = state.caption
    aid = event.agent_id
    p = event.payload

    match event.type:
        case "run.started":
            run_status = "running"
            caption = f"run started: {p.get('goal', '')}"
        case "run.completed":
            run_status = p.get("status", "completed")
            caption = f"run {run_status}"
        case "run.aborted":
            run_status = "aborted"
            caption = f"run aborted: {p.get('reason', '')}"
        case "agent.spawned":
            key = aid or ""
            parent_id = p.get("parent_id")
            agents[key] = AgentState(
                agent_id=key,
                role=p.get("role"),
                parent_id=parent_id,
                status=Status.IDLE,
                zone=Zone.DESK if parent_id else Zone.FLOOR,
                team=p.get("team"),
            )
            caption = f"spawned {key}" + (f" ({p['role']})" if p.get("role") else "")
        case "task.delegated":
            to_agent = p.get("to_agent", "")
            paths.append((aid or "", to_agent))
            if aid in agents:
                agents[aid].status = Status.DELEGATING
                agents[aid].zone = Zone.PATH
            if to_agent in agents:
                agents[to_agent].status = Status.WORKING
            caption = f"{aid} -> {to_agent}: {p.get('task', '')}"
        case "vault.read":
            if aid in agents:
                agents[aid].zone = Zone.LIBRARY
                agents[aid].status = Status.READING
            caption = f"{aid} reads vault: {p.get('ref', '')}"
        case "decision.pending":
            caption = f"decision pending: {p.get('prompt', '')}"
        case "decision.resolved":
            caption = f"decision resolved: {p.get('choice', '')}"
        case "agent.failed":
            if aid in agents:
                agents[aid].status = Status.FAILED
            caption = f"{aid} failed: {p.get('cause', '')}"
        case _:
            # Valid but unhandled event type: advance without crashing (forward-compat).
            caption = event.type

    return SceneState(
        current_seq=event.seq,
        run_status=run_status,
        caption=caption,
        agents=agents,
        paths=paths,
    )


def reduce_all(
    events: Iterable[Event], initial: SceneState | None = None
) -> SceneState:
    """Fold a whole event sequence into a final SceneState."""
    state = initial if initial is not None else initial_state()
    for event in events:
        state = reduce(state, event)
    return state
